### PR TITLE
Smoothes gravity wells

### PR DESCRIPTION
**Pull request by TemporalOroboros**
## About the PR 
- Smooths out the acceleration entities experience when nearby gravity wells.
- Moves gravity wells to shared so they can be predicted again.

**Media**

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

https://user-images.githubusercontent.com/53582584/211671311-871432f0-9c42-4f8a-a454-449b955f72a6.mp4

**Changelog**

:cl:
- add: The singularity and similar objects will now gradually accelerate objects near themselves instead of violently yanking them across the room.

### CODE REVIEW

File: Content.Server/Singularity/Components/GravityWellComponent.cs [hunk 1]
```diff
@@ -43,29 +43,43 @@ public sealed class GravityWellComponent : Component
...
+    public TimeSpan NextScanTime { get; internal set; } = default!;
 
     /// <summary>
-    /// The last time this gravity well pulsed.
+    /// The last time this gravity well scanned for captured entities.
     /// </summary>
     [ViewVariables(VVAccess.ReadOnly)]
     [Access(typeof(GravityWellSystem))]
-    public TimeSpan LastPulseTime { get; internal set; } = default!;
+    public TimeSpan LastScanTime { get; internal set; } = default!;
```
Comment (metalgearsloth) [inline Content.Server/Singularity/Components/GravityWellComponent.cs:82]: Last time isn't necessary if you know the frequency and next times.

File: Content.Server/Singularity/EntitySystems/GravityWellSystem.cs [hunk 2]
```diff
@@ -61,201 +58,226 @@ public override void Update(float frameTime)
...
+            if(!TryComp<PhysicsComponent?>(entity, out var physics)
+            || physics.BodyType == BodyType.Static)
+                continue;
+
+            if(!CanGravPulseAffect(entity))
+                continue;
+            
+            gravWell.Captured.Add(entity);
+        }
     }
```
Comment (metalgearsloth) [inline Content.Server/Singularity/EntitySystems/GravityWellSystem.cs:108]: This method is expensive and should be using entityqueries, also you're not querying for static bodies but checking for them? Ideally this should just be applying a force every tick.

File: Content.Server/Singularity/EntitySystems/GravityWellSystem.cs [hunk 3]
```diff
@@ -61,201 +58,226 @@ public override void Update(float frameTime)
...
     /// </summary>
-    /// <param name="uid">The uid of the gravity well to make pulse.</param>
-    /// <param name="gravWell">The state of the gravity well to make pulse.</param>
-    /// <param name="frameTime">The amount to consider as having passed since the last gravitational pulse by the gravity well. Pulse force scales with this.</param>
-    /// <param name="xform">The transform of the gravity well to make pulse.</param>
+    /// <param name="uid">The uid of the gravity well to make scan.</param>
+    /// <param name="gravWell">The state of the gravity well to make scan.</param>
+    /// <param name="frameTime">The amount to consider as having passed since the last gravitational scan by the gravity well. Scan force scales with this.</param>
+    /// <param name="xform">The transform of the gravity well to make scan.</param>
     private void Update(EntityUid uid, TimeSpan frameTime, GravityWellComponent? gravWell = null, TransformComponent? xform = null)
```
Comment (metalgearsloth) [inline Content.Server/Singularity/EntitySystems/GravityWellSystem.cs:85]: frametime is unused.

File: Content.Server/Singularity/EntitySystems/GravityWellSystem.cs [hunk 4]
```diff
@@ -27,30 +25,29 @@ public sealed class GravityWellSystem : SharedGravityWellSystem
...
         base.Shutdown();
     }
 
     /// <summary>
-    /// Updates the pulse cooldowns of all gravity wells.
-    /// If they are off cooldown it makes them emit a gravitational pulse and reset their cooldown.
+    /// Updates the set of captured entities for all gravity wells.
     /// </summary>
     /// <param name="frameTime">The time elapsed since the last set of updates.</param>
     public override void Update(float frameTime)
```
Comment (metalgearsloth) [inline Content.Server/Singularity/EntitySystems/GravityWellSystem.cs:53]: Looks like someone added IsFirstTimePredicted here but it doesn't even do anything on the server.

File: Content.Server/Singularity/EntitySystems/GravityWellSystem.cs [hunk 5]
```diff
@@ -61,201 +58,226 @@ public override void Update(float frameTime)
...
+    /// If the new scan period implies that the gravity well was intended to scan already it does so immediately.
     /// </summary>
-    /// <param name="uid">The uid of the gravity well to set the pulse period for.</param>
-    /// <param name="value">The new pulse period for the gravity well.</param>
-    /// <param name="gravWell">The state of the gravity well to set the pulse period for.</param>
-    public void SetPulsePeriod(EntityUid uid, TimeSpan value, GravityWellComponent? gravWell = null)
+    /// <param name="uid">The uid of the gravity well to set the scan period for.</param>
+    /// <param name="value">The new scan period for the gravity well.</param>
+    /// <param name="gravWell">The state of the gravity well to set the scan period for.</param>
+    public void SetScanPeriod(EntityUid uid, TimeSpan value, GravityWellComponent? gravWell = null)
```
Comment (metalgearsloth) [inline Content.Server/Singularity/EntitySystems/GravityWellSystem.cs:233]: Does this even need to be settable.

File: Content.Server/Singularity/Components/GravityWellComponent.cs [hunk 6]
```diff
@@ -43,29 +43,43 @@ public sealed class GravityWellComponent : Component
...
+    [ViewVariables(VVAccess.ReadOnly)]
+    public Matrix3 MatrixAcceleration;
+    
+    /// <summary>
+    /// The set of entities being affected by this gravity well.
+    /// All entities within this group are accelerated relative to the gravity well.
+    /// </summary>
+    [ViewVariables(VVAccess.ReadOnly)]
+    public List<EntityUid> Captured = new();
+
```
Comment (metalgearsloth) [inline Content.Server/Singularity/Components/GravityWellComponent.cs:59]: This component should be using access.

File: Content.Server/Singularity/EntitySystems/GravityWellSystem.cs [hunk 7]
```diff
@@ -61,201 +58,226 @@ public override void Update(float frameTime)
...
+        foreach(var entity in gravWell.Captured)
+        {
+            if(!physicsQuery.TryGetComponent(entity, out var physicsBody)
+            || !xformQuery.TryGetComponent(entity, out var entityXform))
+                continue;
+                    
+            var position = entityXform.MapPosition;
+            if (epicenter.MapId != position.MapId)
+                continue;
+
```
Comment (metalgearsloth) [inline Content.Server/Singularity/EntitySystems/GravityWellSystem.cs:173]: This should be re-arranged to be more optimal, checking mapid is free but getting worldposition is expensive for example.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
There's a bunch of unused methods / variables that need deleting and redundant comments (e.g. describing the entityuid).

Also have you profiled this, coz singulo was expensive enough running a couple times a second.

**Comment by TemporalOroboros** [comment]
> There's a bunch of unused methods / variables that need deleting and redundant comments (e.g. describing the entityuid).
> 
> Also have you profiled this, coz singulo was expensive enough running a couple times a second.

I have not and I'm not certain how. I only have VSCode because VS tries to eat my CPU and rider costs $150-$300.

File: Content.Server/Singularity/EntitySystems/GravityWellSystem.cs (hunk 5 above)
Comment (TemporalOroboros) [reply Content.Server/Singularity/EntitySystems/GravityWellSystem.cs:233]: It exists, an admin might want to modify it for <insert reason here>, and doing so will not endanger players so probably.

File: Content.Server/Singularity/EntitySystems/GravityWellSystem.cs (hunk 2 above)
Comment (TemporalOroboros) [reply Content.Server/Singularity/EntitySystems/GravityWellSystem.cs:108]: Ideally yes, unfortunately the physics system clears all forces applied to entities every tick and if I use impulse to scale according to tick time we're back to entities being yanked around.

I could use a component to track entities in range of the singularity, but I was under the impression that the impending Archetypes would make adding/removing components about an order of magnitude more expensive.

**Comment by metalgearsloth** [comment]
> > There's a bunch of unused methods / variables that need deleting and redundant comments (e.g. describing the entityuid).
> > Also have you profiled this, coz singulo was expensive enough running a couple times a second.
> 
> I have not and I'm not certain how. I only have VSCode because VS tries to eat my CPU and rider costs $150-$300.

Long as we keep trygetcomponent down it should probably be okay I think, worst case I'll just use dottrace later.

File: Content.Server/Singularity/EntitySystems/GravityWellSystem.cs (hunk 7 above)
Comment (TemporalOroboros) [reply Content.Server/Singularity/EntitySystems/GravityWellSystem.cs]: I've made some changes. It should be slightly faster now.

File: Content.Server/Singularity/EntitySystems/GravityWellSystem.cs (hunk 3 above)
Comment (TemporalOroboros) [reply Content.Server/Singularity/EntitySystems/GravityWellSystem.cs]: Switched to current time so we only resolve it once. Might be slightly faster.

File: Content.Server/Singularity/Components/GravityWellComponent.cs (hunk 1 above)
Comment (TemporalOroboros) [reply Content.Server/Singularity/Components/GravityWellComponent.cs:82]: It's used for recalculating the next scan time on VV.

File: Content.Server/Singularity/EntitySystems/GravityWellSystem.cs (hunk 4 above)
Comment (TemporalOroboros) [reply Content.Server/Singularity/EntitySystems/GravityWellSystem.cs:55]: Removed. I honestly just cargo-culted it when I was told to switch to target times and then didn't remove it when I found out how it worked.

File: Content.Server/Singularity/Components/GravityWellComponent.cs (hunk 6 above)
Comment (TemporalOroboros) [reply Content.Server/Singularity/Components/GravityWellComponent.cs]: Done.

**Comment by TemporalOroboros** [comment]
So as it turns out kinematic bodies accept forces being applied to them but don't actually use them during physics sim. This creates an inconsistency between `ApplyLinearImpulse()` which does successfully punt such entities around and `ApplyForce()` which does, effectively, nothing. I've put up an engine PR to resolve this, but in the meantime merging this without that PR will remove singularities pulling on mobs.

**Comment by TemporalOroboros** [comment]
Hold on, I can just use impulse with time scaling. It's slightly more expensive, but it doesn't need an engine PR (that sloth rejected anyways).

File: Content.Server/Singularity/EntitySystems/GravityWellSystem.cs (hunk 2 above)
Comment (TemporalOroboros) [reply Content.Server/Singularity/EntitySystems/GravityWellSystem.cs]: Moved to using EntityQueries. That should be fast enough.

File: Content.Server/Singularity/EntitySystems/GravityWellSystem.cs (hunk 4 above)
Comment (TemporalOroboros) [reply Content.Server/Singularity/EntitySystems/GravityWellSystem.cs]: Added back because it's now in shared.

**Comment by AJCM-git** [comment]
reopen once you got time

**Pull request closed without merging**